File: binary-search/python/binary.py
def search(haystack, needle):
    return _search(haystack, needle, 0, len(haystack) -1)

def _search(haystack, needle, min, max):
    if max < min: return None

    midpoint = min + (max - min) // 2
    guess = haystack[midpoint]
    if guess == needle:
        return midpoint
    elif guess < needle:
        return _search(haystack, needle, midpoint + 1, max)
    elif guess > needle:
        return _search(haystack, needle, min, midpoint - 1)

File: binary-search/js/binary.js
export const search = (sortedArray, needle) => {
  function _search(arr, needle, min, max) {
    if (max < min) return -1
    const midpoint = min + Math.floor((max - min) / 2)
    const guess = arr[midpoint]
    if (guess === needle) return midpoint
    else if (guess < needle) return _search(arr, needle, midpoint + 1, max)
    else if (guess > needle) return _search(arr, needle, min, midpoint - 1)
  }

  return _search(sortedArray, needle, 0, sortedArray.length - 1)
}
